### ISSUE TITLE
Install hangs indefinitely when rvm needs to install ruby version not pre-compiled

### ISSUE THREAD

**Issue by gmalouf** [rvm/rvm1-ansible#70]
We are using ruby-2.0.0 on Ubuntu Trust for our ec2 machines.  We've been experiencing a behavior where the ansible deployment hangs indefinitely even after rvm has finished installing ruby as a part of this command:

- name: Install rubies
  command: '{{ rvm1_rvm }} install {{ item.item }} {{ rvm1_ruby_install_flags }}'
  when: rvm1_rubies and item.rc != 0
  with_items: detect_rubies.results
  sudo_user: '{{ rvm1_user }}'

Restarting ansible (after ensuring ruby was installed on the box manually) allows the deployment to continue to the next tasks.

**Comment by nickjj** [comment]
How long did you give it to compile? On lower end instances it could take a long time.

**Comment by gmalouf** [comment]
It takes like 20 minutes to compile - the issue is that ansible just keeps hanging even after this has finished...

**Comment by Careerbuilder2016** [comment]
Hi, I am encountering this problem as well. Installer hangs for very long time on the install rubies step. Any fix for this?

**Comment by nickjj** [comment]
Not that I know of. I haven't been able to reproduce it. Are you also using ruby-2.0.0 specifically @Careerbuilder2016?

**Comment by Careerbuilder2016** [comment]
I am testing the playbook on Ubuntu Trusty Vagrant images. Any ruby hangs for me, but when I log into the VMs, I see RVM and all ruby versions installed even when ansible hangs.

**Comment by Careerbuilder2016** [comment]
Hi,
So it turns out that rvm calls a lot on apt-get. So when you actually run rvm install, apt-get is ran before it and updates the repository sources needed to be updated. If any one of the sources returns a 401 or 404, it breaks. If you force the apt-get update in a task before rvm install, it makes sure that rvm can call on apt-get update and install the dependencies for that ruby version. Also it helps if the target OS version has the rvm binaries. :+1: 

```rvm list remote``` to check if your OS has the rvm binaries.

The work around for me was to add a task in tasks/rubies/yml before the Install rubies task to force apt-get to update so to ensure that no error will happen when ```rvm install``` command runs next:

```
- name: Update apt-get to avoid errors
  command: 'sudo apt-get update'
  sudo_user: '{{ rvm1_user }}'
```

We still have not tested this in Ansible Tower. Maybe we won't need this custom fix there.

**Comment by nickjj** [comment]
This seems out of scope for this role. You should always be running an apt-get update on a fresh server.

**Comment by Careerbuilder2016** [comment]
Yes, but the reason Ansible can hang up on that task is when apt-get update has not been ran before. Therefore, the RVM script does not exit properly when that first apt-get call returns improper output and Ansible will hang as a result of that. 

I did not plan to run apt-get before running this rvm role. For a Ruby Rails server, rvm should manage my commands to apt-get and download my first ruby. Otherwise, what good is rvm to me? Should I just run apt-get and manage my ruby versions myself? That's what rvm is for.

I am not arguing that you should run apt-get update in a separate task before the "install rubies task" like I did, but you should document somewhere that apt-get should run before that step as rvm will hang Ansible.

This is actually a combination of how rvm script handles apt-get, not just an Ansible issue.

Thanks Nick!

**Comment by nickjj** [comment]
Running an `apt-get update` is something you would typically run when spinning up a fresh server, with or without ruby. You absolutely should be running this in a task before using this role, it's a general thing to do.

rvm's use case is to install and manage ruby for you. Either by pulling down pre-compiled binaries, or compiling it for you.

If you feel that rvm should report a proper error for this, then I recommend opening a ticket in the rvm repo because this project just uses rvm under the hood to automate installing/managing ruby with ansible.

**Comment by holms** [comment]
@nickjj i just had this problem already with your role, and i've found what's the problem. rvm want's to install some dependencies, and it does `apt-get update` first, and then install few packages. I've solved this by finding rvm dependencies with `rvm requirements` command (i had to add user to sudoers for this), and then i've got those two packages which rvm tried to install.. and added to the task.. and now it works :)
